### PR TITLE
Generate storage info for pallet im_online

### PR DESCRIPTION
**Pull request by georgesdib**
Fixes part of #8629

Refactor Im Online to use WeakBoundedVec instead of Vec, and generate storage info for that pallet.

polkadot companion: paritytech/polkadot#3744

@shawntabrizi @KiChjang @thiolliere

### CODE REVIEW

File: frame/im-online/src/lib.rs [hunk 1]
```diff
@@ -333,14 +345,21 @@ pub mod pallet {
...
 	#[pallet::getter(fn received_heartbeats)]
-	pub(crate) type ReceivedHeartbeats<T> =
-		StorageDoubleMap<_, Twox64Concat, SessionIndex, Twox64Concat, AuthIndex, Vec<u8>>;
+	pub(crate) type ReceivedHeartbeats<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		SessionIndex,
+		Twox64Concat,
+		AuthIndex,
+		WeakBoundedVec<u8, T::MaxKeys>,
```
Comment (thiolliere) [inline frame/im-online/src/lib.rs:361]: the value here was a `Vec<u8>` which contains the encoded network state (i.e. an encoded value of type `OpaqueNetworkState`.
So it is no bounded by the max number of keys. We have to find or ask for the bound somewhere else.
It seems to be bounded by the max number of external addresses that the node can be reached at. maybe we can bound it directly in `sp-core::offchain::OpaqueNetworkState` or (and it is easier) ask for the bound in a generic param.

File: frame/im-online/src/lib.rs [hunk 2]
```diff
@@ -410,10 +429,13 @@ pub mod pallet {
 				Self::deposit_event(Event::<T>::HeartbeatReceived(public.clone()));
 
 				let network_state = heartbeat.network_state.encode();
+				let network_state_bounded =
+					WeakBoundedVec::<u8, T::MaxKeys>::try_from(network_state)
+						.map_err(|_err| Error::<T>::SizeTooLargeForBoundedVec)?;
```
Comment (thiolliere) [inline frame/im-online/src/lib.rs:434]: here is where we incorrectly expect the network state to be bounded by MaxKeys.

**Review by thiolliere — CHANGES_REQUESTED** [review]

File: frame/im-online/src/lib.rs (hunk 1 above)
Comment (georgesdib) [reply frame/im-online/src/lib.rs:361]: The main reason why I did not bound it in `sp-core::off chain::OpaqueNetworkState` was because `WeakBoundedVec` is in `frame-support` and `sp-core` should not depend on it.

But indeed, you are right, I will have a think as to what is the best option to approach this, and will suggest it. Thanks!

Comment (georgesdib) [reply frame/im-online/src/lib.rs:361]: Question: `ReceivedHeartbeats` stores the encoding of the network state, but never seem to actually use that value, what is the purpose of storing that? Can we store its hash instead and that would naturally be bounded?

Comment (georgesdib) [reply frame/im-online/src/lib.rs]: Failing to find a better alternative, I have added MaxPeerInHeartbeats to bound ReceivedHeartbeats. It's a WeakBoundedVec and does not fail if the number is too big, but emits a warning. Let me know if this looks good to you.

Comment (thiolliere) [reply frame/im-online/src/lib.rs]: > Question: ReceivedHeartbeats stores the encoding of the network state, but never seem to actually use that value, what is the purpose of storing that? Can we store its hash instead and that would naturally be bounded?

Maybe it is only stored on chain so that it can be read, or can be used for some proof later, but I don't know either.

> Failing to find a better alternative, I have added MaxPeerInHeartbeats to bound ReceivedHeartbeats. It's a WeakBoundedVec and does not fail if the number is too big, but emits a warning. Let me know if this looks good to you.

this seems a good solution to me.

File: frame/im-online/src/lib.rs [hunk 3]
```diff
@@ -260,7 +266,14 @@ pub mod pallet {
...
 			+ Ord
-			+ MaybeSerializeDeserialize;
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen;
+
+		/// The maximum number of keys that can be added.
+		type MaxKeys: Get<u32>;
+
+		/// The maximum number of peers to be stored in ReceivedHeartbeats
+		type MaxPeerInHeartbeats: Get<u32>;
```
Comment (thiolliere) [inline frame/im-online/src/lib.rs:276]: This doesn't describe its current usage. Currently this limit the number of bytes the encoding of the network state of the heartbeat.
Not the number of Peer.

Either we update the doc to say precisely what is this limit. But then this limit is a bit difficult to configure, or we update our usage of this limit.

We can do that by create a structure `BoundedHeartBeat<BlockNumber, Limit>` with the field `BoundedOpaqueNetworkState<Limit>` which would be generic over the limit. Maybe we can create a wrapper type to encode BoundedHeartBeat with the compact u32 in front of type. like a new struct `WrapperOpaque<Foo>(Foo)`.

So then we implement MaxEncodedLen on WrapperOpaque, BoundedHeartBeat and BoundedOpaqueNetworkState. And put into the storage the type `WrapperOpaque<BoundedHeartBeat<BlockNumber, MaxPeerInHearbeats>>`

(the type WrapperOpaque might be handy we can put it in frame_support probably)

Comment (georgesdib) [reply frame/im-online/src/lib.rs:276]: Good point, why do we need `BoundedHeartbeat` though?
Should we not be storing `WrapperOpaque<BoundedOpaqueNetworkState<_,_>>` instead?
Unless you are suggesting to change the signature of the extrinsic `heartbeat` but that is a bit of an overkill, no?

Comment (thiolliere) [reply frame/im-online/src/lib.rs:276]: Yes I mix myself, `WrapperOpaque<BoundedOpaqueNetworkState<_,_>>` is what we want indeed

Comment (georgesdib) [reply frame/im-online/src/lib.rs:276]: I am not too sure to understand what do you mean by WrapperOpaque? What is its purpose? Why not store directly `BoundedOpaqueNetwork` ?

Comment (thiolliere) [reply frame/im-online/src/lib.rs:276]: Yes WrapperOpaque was just a helper struct, in order not to change the encoding, (because currently it is encoded as a `Vec<u8>`).
I introduce this struct in this PR https://github.com/paritytech/substrate/pull/9738
So you could use `WrapperOpaque<BoundedNetworkState<..>>` where BoundedNetworkState is using a simple `#[derive(Encode,Decode)]` or `BoundedOpaqueNetworkState` but then you need to implement the encoding/decoding manually I think.

Comment (georgesdib) [reply frame/im-online/src/lib.rs:341]: Oh great, I will take a look and amend accordingly. It was a bit tricky this refactoring.
In particular, if I wanted to avoid adding some dependencies, PeerId should in fact have a bounded encoding, but it's difficult to add at source. MultiAddr could in theory be unbounded, but in practice would be, so I added one extra variable MaxPeerIdEncodingSize which represents the max encoding of these 2 variables. Please let me know if what I did is in the right track, and I will also refactor to include your inclusion of `WrapperOpaque`
Thanks!

Comment (georgesdib) [reply frame/im-online/src/lib.rs:331]: I have pushed through your addition of `WrapperOpaque` adding a few things to integrate it. Obviously, happy if you prefer to wait for your PR to go live first if you prefer.

File: frame/im-online/src/lib.rs [hunk 4]
```diff
@@ -219,6 +227,58 @@ where
 	pub validators_len: u32,
 }
 
+/// A type that is the same as [`OpaqueNetworkState`] but with [`Vec`] replaced with
+/// [`WeakBoundedVec<Limit>`] where Limit is the respective size limit
+#[derive(Clone, Eq, PartialEq, Encode, Decode, MaxEncodedLen)]
+#[codec(mel_bound(EncodingLimit: Get<u32>, AddressesLimit: Get<u32>))]
+pub struct BoundedOpaqueNetworkState<EncodingLimit, AddressesLimit>
```
Comment (thiolliere) [inline frame/im-online/src/lib.rs:234]: note that you use here `EncodingLimit` for both the limit of `OpaquePeerId` and the limit of `OpaqueMultiAddr`.
Maybe we should also differenciate both. or at least put in the doc that this limit should be the max of both limits.

**Review by thiolliere — COMMENTED** [review]
apart from this last note, I think it should be good to me

File: frame/im-online/src/lib.rs (hunk 3 above)
Comment (thiolliere) [reply frame/im-online/src/lib.rs:331]: I think it is good to have it here so we can merge this PR first.

File: frame/im-online/src/lib.rs (hunk 4 above)
Comment (georgesdib) [reply frame/im-online/src/lib.rs]: I have committed changes to address that by differentiating the limits.

File: frame/support/src/traits/misc.rs [hunk 5]
```diff
@@ -377,3 +379,48 @@ impl<Call, Balance: From<u32>, const T: u32> EstimateCallFee<Call, Balance> for
...
+/// Thus the encoded value can be decoded as a `Vec<u8>`.
+#[derive(Debug, Eq, PartialEq, Default, Clone, MaxEncodedLen)]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+pub struct WrapperOpaque<T>(pub T);
+
+impl<T: Encode> EncodeLike for WrapperOpaque<T> {}
+
+impl<T: Encode> Encode for WrapperOpaque<T> {
+	fn size_hint(&self) -> usize {
+		// Compact<u32> usually takes less than 4 bytes
```
Comment (thiolliere) [inline frame/support/src/traits/misc.rs:395]:
```suggestion
		// Compact<u32> usually takes at most 4 bytes
```

**Review by thiolliere — APPROVED** [review]
looks good to me, maybe name can be improved slightly

File: frame/im-online/src/lib.rs [hunk 6]
```diff
@@ -260,7 +327,18 @@ pub mod pallet {
...
+
+		/// The maximum number of keys that can be added.
+		type MaxKeys: Get<u32>;
+
+		/// The maximum number of peers to be stored in ReceivedHeartbeats
+		type MaxPeerInHeartbeats: Get<u32>;
+
+		/// The maximum size of the encoding of PeerId and MultiAddr that are coming
+		/// from the hearbeat
+		type MaxPeerIdEncodingSize: Get<u32>;
```
[the remark this answers is not in the document]
Comment (georgesdib) [reply frame/im-online/src/lib.rs]: Done

[the remark this answers is not in the document]
Comment (georgesdib) [reply frame/im-online/src/lib.rs]: Do you mind tagging who the relevant second reviewer should be? Sorry new to this repo, thanks.

File: frame/im-online/src/lib.rs [hunk 7]
```diff
@@ -74,9 +74,14 @@ mod mock;
...
 pub mod weights;
 
-use codec::{Decode, Encode};
-use frame_support::traits::{
-	EstimateNextSessionRotation, OneSessionHandler, ValidatorSet, ValidatorSetWithIdentification,
+use codec::{Decode, Encode, MaxEncodedLen};
+use core::convert::TryFrom;
+use frame_support::{
+	traits::{
+		EstimateNextSessionRotation, OneSessionHandler, ValidatorSet,
```
Comment (KiChjang) [inline frame/im-online/src/lib.rs:81]:
```suggestion
		EstimateNextSessionRotation, Get, OneSessionHandler, ValidatorSet,
```

File: frame/im-online/src/lib.rs [hunk 8]
```diff
@@ -92,6 +97,9 @@ use sp_staking::{
 	SessionIndex,
 };
 use sp_std::{convert::TryInto, prelude::*};
+
+use frame_support::traits::Get;
```
Comment (KiChjang) [inline frame/im-online/src/lib.rs:101]: Remove this.

File: frame/im-online/src/lib.rs [hunk 9]
```diff
@@ -736,16 +831,20 @@ impl<T: Config> Pallet<T> {
 		res
 	}
 
-	fn initialize_keys(keys: &[T::AuthorityId]) {
+	fn initialize_keys(keys: &Vec<T::AuthorityId>) {
```
Comment (KiChjang) [inline frame/im-online/src/lib.rs:834]: This is a stricter requirement than `&[T::AuthorityId]`, why the change?

File: frame/im-online/src/lib.rs (hunk 7 above)
Comment (georgesdib) [reply frame/im-online/src/lib.rs]: Good catch, amended.

File: frame/im-online/src/lib.rs (hunk 9 above)
Comment (georgesdib) [reply frame/im-online/src/lib.rs]: Can't remember why, but changing it back to a slice works, so I have changed it, thanks

File: frame/im-online/src/lib.rs (hunk 8 above)
Comment (georgesdib) [reply frame/im-online/src/lib.rs]: Removed

**Review by KiChjang — APPROVED** [review]

**Comment by KiChjang** [comment]
CI is complaining:

```
error[E0277]: the trait bound `frame_benchmarking::Vec<<T as pallet::Config>::AuthorityId>: EncodeLike<frame_support::WeakBoundedVec<<T as pallet::Config>::AuthorityId, <T as pallet::Config>::MaxKeys>>` is not satisfied
  --> frame/im-online/src/benchmarking.rs:49:17
   |
49 |     Keys::<T>::put(keys.clone());
   |                    ^^^^^^^^^^^^ the trait `EncodeLike<frame_support::WeakBoundedVec<<T as pallet::Config>::AuthorityId, <T as pallet::Config>::MaxKeys>>` is not implemented for `frame_benchmarking::Vec<<T as pallet::Config>::AuthorityId>`
   |
   = help: the following implementations were found:
             <frame_benchmarking::Vec<T> as EncodeLike<&[U]>>
             <frame_benchmarking::Vec<T> as EncodeLike<VecDeque<U>>>
             <frame_benchmarking::Vec<T> as EncodeLike<frame_benchmarking::Vec<U>>>
   = note: required by `frame_support::pallet_prelude::StorageValue::<Prefix, Value, QueryKind, OnEmpty>::put`
```

**Comment by georgesdib** [comment]
Rebased my branch to make it cleaner and force the checks of the polkadot companion

**Comment by thiolliere** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge
